### DISCUSSION TITLE
Notice X01 issue

### DISCUSSION THREAD

**Discussion by rengit62** [OP-TED/eForms-SDK#465] [Q&A] [answered]
Hi Lab
in notice X01 , can you tell me where is  the field id of AdditionalDocumentReference ? 
In front-end form I see 4 fields correspondent to : 
		OPP-120-Business[repeatable=false,xpathRelative=cbc:DocumentDescription]
		OPP-121-Business[repeatable=false,xpathRelative=cbc:ReferencedDocumentInternalAddress]
		OPP-122-Business[repeatable=false,xpathRelative=cac:Attachment/cac:ExternalReference/cbc:URI]
		OPP-123-Business[repeatable=false,xpathRelative=cbc:IssueDate]

![image](https://github.com/OP-TED/eForms-SDK/assets/52211037/4adc61ee-a16e-42c4-ad4f-5b0149cf0e64)

in XSD id is mandatory : 

` <xsd:complexType name="DocumentReferenceType">
      <xsd:sequence>
         <xsd:element ref="ext:UBLExtensions" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:ID" minOccurs="1" maxOccurs="1"/>
         <xsd:element ref="cbc:CopyIndicator" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:UUID" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:IssueDate" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:IssueTime" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:DocumentTypeCode" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:DocumentType" minOccurs="0" maxOccurs="unbounded"/>
         <xsd:element ref="cbc:XPath" minOccurs="0" maxOccurs="unbounded"/>
         <xsd:element ref="cbc:ReferencedDocumentInternalAddress"
                      minOccurs="0"
                      maxOccurs="1"/>
         <xsd:element ref="cbc:LanguageID" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:LocaleCode" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:VersionID" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:DocumentStatusCode" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cbc:DocumentDescription" minOccurs="0" maxOccurs="unbounded"/>
         <xsd:element ref="cac:Attachment" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cac:ValidityPeriod" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cac:IssuerParty" minOccurs="0" maxOccurs="1"/>
         <xsd:element ref="cac:ResultOfVerification" minOccurs="0" maxOccurs="1"/>
      </xsd:sequence>
   </xsd:complexType>
  `

Thanks and Regards
Renato

**Comment by YvesJo** [comment]
This is an aggregate component (cac:AdditionalDocumentReference) which is associated with the display group Publication containing the 4 fields (Name of Publication, Title of the announcement, URL of the announcement, Date of publication).

**Reply by rengit62** [reply]
What does it mean "an aggregate component "  ? 
In SDK  Examples notices, there is notice  X01_EEIG.xml and under cac:AdditionalDocumentReference there is : 
<cbc:ID>DUMMY</cbc:ID>

When the app finds AdditionalDocumentReference should add a new field called cbc:ID with value DUMMY ?
Please can you explain it ?
Thanks in advance
Renato

**Reply by YvesJo** [reply]
cac: common aggregate component, i.e. that further contains elements (branch)
cbc: common basic component
What you're actually referring to is the generation of the cbc:ID which is mandatory according to the schema, So when generating the XML, indeed an element cbc:ID should automatically be added in order to comply with it. There is no BT or field associated to it and it shouldn't convey additional information.

**Reply by rengit62** [reply]
I know that the  cbc:ID  is mandatory according to the schema, what I asked you how a java program can understand, from SDK,  that it must add a  cbc:id ( with value DUMMY ).
in X01.json those 4 fields are children of  group GR-Business-Publication which has not got a node : 
"id" : "GR-Business-Publication",
      "contentType" : "group",
      "displayType" : "GROUP",
      "description" : "Publication",
      "_label" : "group|name|GR-Business-Publication",
      "content" : [ {
        "id" : "OPP-120-Business",
        "contentType" : "field",
        "displayType" : "TEXTAREA",
        "description" : "Name of Publication",
        "_label" : "field|name|OPP-120-Business"
      }, {
        "id" : "OPP-121-Business",
        "contentType" : "field",
        "displayType" : "TEXTAREA",
        "description" : "Title of the announcement",
        "_label" : "field|name|OPP-121-Business"
      }, {
        "id" : "OPP-122-Business",
        "contentType" : "field",
        "displayType" : "TEXTBOX",
        "description" : "URL of the announcement",
        "_label" : "field|name|OPP-122-Business"
      }, {
        "id" : "OPP-123-Business",
        "contentType" : "field",
        "displayType" : "TEXTBOX",
        "description" : "Date of publication",
        "_label" : "field|name|OPP-123-Business"
      }

in fields.json the same fields are children of ND-GazetteReference
ND-GazetteReference[repeatable=false,xpathRelative=cac:AdditionalDocumentReference]
		OPP-120-Business[repeatable=false,xpathRelative=cbc:DocumentDescription]
		OPP-121-Business[repeatable=false,xpathRelative=cbc:ReferencedDocumentInternalAddress]
		OPP-122-Business[repeatable=false,xpathRelative=cac:Attachment/cac:ExternalReference/cbc:URI]
		OPP-123-Business[repeatable=false,xpathRelative=cbc:IssueDate]

cac:AdditionalDocumentReference is of type DocumentReferenceType in which cbc:ID, OK.
 <xsd:element ref="cbc:ID" minOccurs="1" maxOccurs="1"/>

Writing "should automatically be added in order to comply with it" do you mean that ALWAYS , when it lacks a field cbc:ID it must be added with a Dummy value ? Or only for tag of type DocumentReferenceType ? Only in notice X01 ?

Thanks a lot !
Renato

**Comment by rouschr** [comment, marked as answer]
> add a new field called cbc:ID with value DUMMY ?

If cbc:ID is mandatory (`<xsd:element ref="cbc:ID" minOccurs="1" maxOccurs="1"/>`)
Then you need to add it, the value you put into the form field may also be restricted by a pattern, but in general mandatory means some value must be put inside of it.
The example DUMMY value is just one possible that was put there in order to fill the mandatory field.
As long as the pattern does not enforce something you are free to put some values there (user filled, machine filled, ...).

Example of a pattern would be:
```
 "pattern" : {
      "value" : "^UBO-\\d{4}$",
      "severity" : "ERROR"
    },
```

And in some cases you even have this to help you with ID generation:
```
   "type" : "id",
    "idScheme" : "UBO",
    "schemeName" : "ubo",
```

You can find some documentation about this here:
https://docs.ted.europa.eu/eforms/latest/fields/index.html

**Reply by rengit62** [reply]
OK thanks, I added that field.
Now I have not got more xsd error , but 3 validation errors : 

**********************************
id : BR-BT-00501-0051
location : /BusinessRegistrationInformationNotice/cac:BusinessParty
failedAssertText : BT-501-Business-National is not allowed in notice type 'X01'
diagnostic : ND-BusinessParty_BT-501-Business-National
see : field:BT-501-Business-National
diagnoticReferenceText : cac:PartyLegalEntity[not(cbc:CompanyID/@schemeName = 'EU')]/cbc:CompanyID[not(@schemeName = 'EU')]


**********************************
id : BR-BT-00501-0207
location : /BusinessRegistrationInformationNotice/cac:BusinessParty
failedAssertText : BT-501-Business-European is mandatory in notice type 'X01'
diagnostic : ND-BusinessParty_BT-501-Business-European
see : field:BT-501-Business-European
diagnoticReferenceText : cac:PartyLegalEntity[cbc:CompanyID/@schemeName = 'EU']/cbc:CompanyID[@schemeName = 'EU']


**********************************
id : BR-BT-00500-0156
location : /BusinessRegistrationInformationNotice/cac:BusinessParty/cac:PartyLegalEntity[2]
failedAssertText : BT-500-Business is mandatory in notice type 'X01'
diagnostic : BT-500-Business
see : field:BT-500-Business
diagnoticReferenceText : cbc:RegistrationName

the xml I make : 
`<?xml version="1.0" encoding="UTF-8"?>
<BusinessRegistrationInformationNotice xmlns="http://data.europa.eu/p27/eforms-business-registration-information-notice/1" xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2" xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2" xmlns:efac="http://data.europa.eu/p27/eforms-ubl-extension-aggregate-components/1" xmlns:efbc="http://data.europa.eu/p27/eforms-ubl-extension-basic-components/1" xmlns:efext="http://data.europa.eu/p27/eforms-ubl-extensions/1" xmlns:ext="urn:oasis:names:specification:ubl:schema:xsd:CommonExtensionComponents-2">
	<ext:UBLExtensions>
		<ext:UBLExtension>
			<ext:ExtensionContent>
				<efext:EformsExtension>
					<efac:NoticeSubType>
						<cbc:SubTypeCode listName="notice-subtype">X01</cbc:SubTypeCode>
					</efac:NoticeSubType>
				</efext:EformsExtension>
			</ext:ExtensionContent>
		</ext:UBLExtension>
	</ext:UBLExtensions>
	<cbc:UBLVersionID>2.3</cbc:UBLVersionID>
	<cbc:CustomizationID>eforms-sdk-1.7</cbc:CustomizationID>
	<cbc:ID schemeName="notice-id">91395664-89ed-46b4-8dd1-6d97e87cad46</cbc:ID>
	<cbc:IssueDate>2023-06-07+02:00</cbc:IssueDate>
	<cbc:IssueTime>02:45:52+02:00</cbc:IssueTime>
	<cbc:VersionID>01</cbc:VersionID>
	<cbc:RegulatoryDomain>31985R2137</cbc:RegulatoryDomain>
	<cbc:NoticeTypeCode listName="bri">brin-eeig</cbc:NoticeTypeCode>
	<cbc:NoticeLanguageCode listName="language">ENG</cbc:NoticeLanguageCode>
	<cbc:Note languageID="ENG">IN SU LA VETTA DELLA TORRE ANTICA</cbc:Note>
	<cac:SenderParty>
		<cac:Contact>
			<cbc:ElectronicMail>ferdi@gmail.sd</cbc:ElectronicMail>
		</cac:Contact>
	</cac:SenderParty>
	<cac:BusinessParty>
		<cac:PartyLegalEntity>
			<cbc:RegistrationName>FERDINANDO MARIO PIO e GERMANI</cbc:RegistrationName>
		</cac:PartyLegalEntity>
		<cac:PartyLegalEntity>
			<cbc:CompanyID schemeName="'EU">1</cbc:CompanyID>
			<cac:CorporateRegistrationScheme>
				<cac:JurisdictionRegionAddress>
					<cbc:CityName>PISA</cbc:CityName>
					<cbc:PostalZone>40123</cbc:PostalZone>
					<cac:Country>
						<cbc:IdentificationCode listName="country">ITA</cbc:IdentificationCode>
					</cac:Country>
				</cac:JurisdictionRegionAddress>
			</cac:CorporateRegistrationScheme>
		</cac:PartyLegalEntity>
	</cac:BusinessParty>
	<cac:AdditionalDocumentReference>
		<cbc:ID>DUMMY</cbc:ID>
		<cbc:IssueDate>2023-05-31+02:00</cbc:IssueDate>
		<cbc:ReferencedDocumentInternalAddress>GEPPETTO e PINOCCHIO</cbc:ReferencedDocumentInternalAddress>
		<cbc:DocumentDescription>GEPPETTO</cbc:DocumentDescription>
		<cac:Attachment>
			<cac:ExternalReference>
				<cbc:URI>http://localh.er</cbc:URI>
			</cac:ExternalReference>
		</cac:Attachment>
	</cac:AdditionalDocumentReference>
	<efac:NoticePurpose>
		<cbc:PurposeCode listName="notice-purpose">reg</cbc:PurposeCode>
	</efac:NoticePurpose>
</BusinessRegistrationInformationNotice>`

Looking at the sample X01_EEIG.xml I guess that the problem how to manage these fields : 
ND-LocalEntity[repeatable=false,xpathRelative=cac:PartyLegalEntity[not(cbc:CompanyID/@schemeName = 'EU')]]
			BT-500-Business[repeatable=false,xpathRelative=cbc:RegistrationName]
		ND-EuEntity[repeatable=false,xpathRelative=cac:PartyLegalEntity[cbc:CompanyID/@schemeName = 'EU']]
			BT-501-Business-European[repeatable=false,xpathRelative=cbc:CompanyID[@schemeName = 'EU']]

I seems that ND-LocalEntity should be discarded but which is the 'logic' ? I mean there is a predicate [not(cbc:CompanyID/@schemeName = 'EU')] but to what it does refers to ?
Can you help me ?
Thanks again and Regards
Renato

**Reply by YvesJo** [reply]
Hi,
There is an unexpected character in the CompanyID schemeName attribute value.

**Reply by rouschr** [reply]
As Yves said:  `<cbc:CompanyID schemeName="'EU">` you have single quote EU, instead of EU, hard to spot :)
